scheduler/proxy/proxy.go

```diff
@@ -0,0 +1,49 @@
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"gocloud.dev/pubsub"
+)
+
+type MessageMutateFunc func(*pubsub.Message) (*pubsub.Message, error)
+
+type PubSubProxy struct {
+	topic        *pubsub.Topic
+	subscription *pubsub.Subscription
+}
+
+func New(topic *pubsub.Topic, subscription *pubsub.Subscription) *PubSubProxy {
+	return &PubSubProxy{
+		topic:        topic,
+		subscription: subscription,
+	}
+}
+
+func (proxy *PubSubProxy) Listen(ctx context.Context, preprocess MessageMutateFunc) error {
+	for {
+		msg, err := proxy.subscription.Receive(ctx)
+		if err != nil {
+			log.Println("error receiving message: ", err)
+			continue
+		}
+		go func(m *pubsub.Message) {
+			outMsg, err := preprocess(msg)
+			if err != nil {
+				log.Println("error processing message: ", err)
+				msg.Nack()
```

Comment:
Rather than nacking in these cases, it's better to rely on the subscription's dead lettering to handle dropping messages on X attempts to delivery. Pub/Sub supports this, but I'm assuming Kafka also supports this? 

This way, if we have a transient error of some sort, we retry a few times rather than dropping the message completely.

Reply:
Although, I think this is still incorrect as the docs state

> Nack is a performance optimization for retrying transient failures. It must not be used for message parse errors or other messages that the application will never be able to process